### ISSUE TITLE
Why i got precision=1 on recommend for topN

### ISSUE THREAD

**Issue by catmango** [NicolasHug/Surprise#61]
I got precision=1 on recommend for topN, that means all predict recommends are all in the user ratings test set，here is my code, thanks for providing the Surprise lib.
```python
user_to_cars = {}
with open('./rating.csv') as file:
    for line in file:
        fields = line.split(',')
        if fields[0] not in user_to_cars:
            user_to_cars[fields[0]] = []
        user_to_cars[fields[0]].append(fields[1])

def main():
    reader = surprise.Reader(line_format='user item rating', sep=',')
    data   = surprise.Dataset.load_from_file('./rating.csv', reader=reader)

    raw_ratings = data.raw_ratings
    #random.shuffle(raw_ratings)
    threshold = int(.8 * len(raw_ratings))
    A_raw_ratings = raw_ratings[:threshold]
    B_raw_ratings = raw_ratings[threshold:]
    data.raw_ratings = A_raw_ratings
    data.split(n_folds=5)

    print('grid search....', arrow.now())
    param_grid = {
        'n_epochs': [5, 10],
        'reg_u': [12, 15],
        'reg_i': [1, 2],
        }
    #param_grid = OrderedDict(sorted(param_grid.items(), key=lambda x: x[0]))
    #param_names  = list(param_grid.keys())
    #param_values = param_grid.values()
    #conbinations = list(itertools.product(*param_values))
    #conbinations = list(map(lambda x: {param_names[i]:x[i] for i in range(len(x))}, conbinations))

    #############
    # grid_search = surprise.GridSearch(ALS, param_grid, measures=['RMSE'], verbose=1)
    # grid_search.evaluate(data)
    # model = grid_search.best_estimator['RMSE']
    ########

    model    = surprise.SVD()
    trainset = data.build_full_trainset()
    model.train(trainset)

    predictions = model.test(trainset.build_testset())
    print('Biased accuracy on A,', end='   ')
    surprise.accuracy.rmse(predictions)

    testset = data.construct_testset(B_raw_ratings)  # testset is now the set B
    predictions = model.test(testset)
    print('Unbiased accuracy on B,', end=' ')
    surprise.accuracy.rmse(predictions)

    top_n = get_top_n(predictions, 10)
    predict_user_to_cars = {}
    for uid, user_ratings in top_n.items():
        predict_user_to_cars[uid] = [iid for (iid, _) in user_ratings]

    TP = 0
    FN = 0
    FP = 0
    for user, cars in predict_user_to_cars.items():
        for car in cars:
            if car in user_to_cars.get(user):
                TP += 1
            else:
                FP += 1
        FN += len(set(user_to_cars.get(user)) - set(cars))
    print('RECALL', float(TP)/(TP + FN))
    print('PRECISION', float(TP)/(TP + FP))


    # neighbors = model.get_neighbors(1612644, k=5)
    # for movie in neighbors:
    #     print(movie)


class ALS(surprise.BaselineOnly):
    """docstring for ALS"""
    def __init__(self, n_epochs, reg_u, reg_i):
        super(ALS, self).__init__(bsl_options = {'method': 'als',
                                                 'n_epochs': n_epochs,
                                                 'reg_u': reg_u,
                                                 'reg_i': reg_i
                                                  })

def get_top_n(predictions, n=10):
    '''Return the top-N recommendation for each user from a set of predictions.

    Args:
        predictions(list of Prediction objects): The list of predictions, as
            returned by the test method of an algorithm.
        n(int): The number of recommendation to output for each user. Default
            is 10.

    Returns:
    A dict where keys are user (raw) ids and values are lists of tuples:
        [(raw item id, rating estimation), ...] of size n.
    '''

    # First map the predictions to each user.
    top_n = defaultdict(list)
    for uid, iid, true_r, est, _ in predictions:
        top_n[uid].append((iid, est))

    # Then sort the predictions for each user and retrieve the k highest ones.
    for uid, user_ratings in top_n.items():
        user_ratings.sort(key=lambda x: x[1], reverse=True)
        top_n[uid] = user_ratings[:n]

    return top_n


if __name__ == '__main__':
    main()
```

here is my data and output
data:
```
728713,1686164,2
728713,1716390,2
728713,1723086,2
728713,1747391,2
728713,1666812,2
728713,1710628,2
728713,1716979,2
728713,1734785,2
728713,1706818,2
728713,1708028,2
728713,1710331,2
728713,1717235,2
728713,1723364,2
728713,1729926,2
```
output:
```
$ python superise_recommend.py
grid search.... 2017-06-30T16:48:37.460139+08:00
Biased accuracy on A,   RMSE: 3.3885
Unbiased accuracy on B, RMSE: 3.4043
RECALL 0.40555523386336606
PRECISION 1.0
```
